Write down implicit rules for import statements in the code base (#5909)
datalad/datalad: docs/source/design/python_imports.rst

.. -*- mode: rst -*-
.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:

.. _chap_design_python_imports:

************************
Python import statements
************************

.. topic:: Specification scope and status

   This specification describes the current (albeit incomplete) implementation.

The following rules apply to any ``import`` statement in the code base:

> - All imports *must* be absolute.

Reviewer: Draft: yet to finish.

As many (if not all) categorical decisions like this, they tend to need exceptions to the rule to make them pragmatic and efficient.

Formatting aside, here is a suggestion:

```suggestion
- All imports *must* be absolute unless modules are *semantically bound* in the hierarchy.

Instances of *semantically bound* modules:

- Tests: `test_<submodule>.py` should use relative form `from ..submodule import ...` to import functionality subject to testing.

Rationale: to facilitate selective testing etc, tests for functionality in the `submodule` reside under `submodule.tests`. Use of relative imports here binds together location of the functionality and the tests allowing for 1). a corresponding module and its tests to be moved within codebase without requiring any code change; 2). easy and unambiguous identification of what functionality is tested without doing visual names matching.

Note: even though `datalad.tests.utils` module resides under `tests`, absolute import should be used to import it for unless in the `datalad.tests.test_tests_utils` which tests functionality of that module.

- Import of base interfaces or specific implementations within some interface: ... 
```

Author: re tests: good point, thx! I did not think much of the tests (in often keep them unchanged in my edits), but it makes sense to handle them here. I am curious what other scenarios of "semantically bound" you will list.

Reviewer: extended (yet to be contracted, added examples to support) above... 

<details>
<summary>here is an example how it would look e.g. in datalad/downloaders/providers.py</summary> 

```python
import os
import re
from collections import OrderedDict
from glob import glob
from logging import getLogger
from os.path import dirname, abspath, join as pathjoin
from urllib.parse import urlparse

from datalad.interface.common_cfg import dirs
from datalad.support.configparserinc import SafeConfigParserWithIncludes
from datalad.support.external_versions import external_versions
from datalad.support.network import RI
from datalad.support import path
from datalad.utils import auto_repr
from datalad.utils import ensure_list_from_str
from datalad.utils import get_dataset_root

from .base import NoneAuthenticator, NotImplementedAuthenticator
from .credentials import CREDENTIAL_TYPES
from .http import (
    HTMLFormAuthenticator,
    HTTPAnonBearerTokenAuthenticator,
    HTTPAuthAuthenticator,
    HTTPBasicAuthAuthenticator,
    HTTPDigestAuthAuthenticator,
    HTTPBearerTokenAuthenticator,
    HTTPDownloader,
)
from .s3 import (
    S3Authenticator,
    S3Downloader,
)
from .shub import SHubDownloader

```
</details>

Reviewer: Just FTR, an obnoxious unification helper for a part of it
```
re='from \.\.\.*\(cmdline\|dochelpers\|downloaders\|support\|ui\)'; git grep -l "$re" | grep -v 'datalad/__init__.py' | xargs sed -i -e "s,$re, from datalad.\1,g"
```

note: any unification we do - we better do it shortly before 0.15.0 release.

Reviewer: <details>
<summary>output of `python -m isort -m3 -fgw 2 -tc datalad/downloaders/providers.py; git diff` on that example I gave above makes it perfect:</summary> 

```diff
diff --git a/datalad/downloaders/providers.py b/datalad/downloaders/providers.py
index b0554cb45..d42b547d8 100644
--- a/datalad/downloaders/providers.py
+++ b/datalad/downloaders/providers.py
@@ -14,27 +14,36 @@ import re
 from collections import OrderedDict
 from glob import glob
 from logging import getLogger
-from os.path import dirname, abspath, join as pathjoin
+from os.path import (
+    abspath,
+    dirname,
+)
+from os.path import join as pathjoin
 from urllib.parse import urlparse
 
 from datalad.interface.common_cfg import dirs
+from datalad.support import path
 from datalad.support.configparserinc import SafeConfigParserWithIncludes
 from datalad.support.external_versions import external_versions
 from datalad.support.network import RI
-from datalad.support import path
-from datalad.utils import auto_repr
-from datalad.utils import ensure_list_from_str
-from datalad.utils import get_dataset_root
+from datalad.utils import (
+    auto_repr,
+    ensure_list_from_str,
+    get_dataset_root,
+)
 
-from .base import NoneAuthenticator, NotImplementedAuthenticator
+from .base import (
+    NoneAuthenticator,
+    NotImplementedAuthenticator,
+)
 from .credentials import CREDENTIAL_TYPES
 from .http import (
     HTMLFormAuthenticator,
     HTTPAnonBearerTokenAuthenticator,
     HTTPAuthAuthenticator,
     HTTPBasicAuthAuthenticator,
-    HTTPDigestAuthAuthenticator,
     HTTPBearerTokenAuthenticator,
+    HTTPDigestAuthAuthenticator,
     HTTPDownloader,
 )
 from .s3 import (
@@ -43,7 +52,6 @@ from .s3 import (
 )
 from .shub import SHubDownloader
 
-
 lgr = getLogger('datalad.downloaders.providers')
 
 # dict to bind authentication_type's to authenticator classes

```
</details>

Author: OK, I can easily get behind the test-scenario. If code is moved, tests will move too. I have done that dozens of times.

I do not get behind the other two scenarios. In particular the annex/gitrepo examples. When I created the recent Repo base classes, I had to RF exactly this kind of "bound" relationship. They are still "bound", by means of subclasses, but not as a monolithic unit.

I think the way I look at it is this:

> Relative imports are desirable/OK, if and only if code is split into multiple files for some technical reason, but semantically belongs to the same place (e.g., code and its tests), i.e. they could life in the same file.

For Git v AnnexRepo it is *one* possiblity to consider them a bound tandem of base classes, but it is just as meaningful to strictly separate annex-related code into a different part of an alternative source tree.

That being said, if we can agree on a short statement, I think that it could be interpreted in a way that also covers the other two scenarios you put forth. I would leave it to the respective developer what is considered a single semantic unit, and where to use relative imports. And I would grant the same freedom to the developer that breaks this unit when moving code and converts the respective imports to absolute ones ;-)

So what about this:

```suggestion
- All imports *must* be absolute, unless they import individual pieces of an integrated code component that is only split across several source code files for technical or organizational reasons.
```

I think this conditions covers all three scenarios you listed.

It is useful to remind ourselves at this point that the aim of these design documents is to provide guidance and practical advise to make it easy to do the "right thing", not to develop enforceable laws ;-)

Reviewer 2: 1. 
I agree, that "the other scenarios" are at the very least less convincing, than the test scenario. Any refactoring can and often will break those relationships of locations, making the relative imports more, not less troublesome in my view, since you'd need to figure out which ones would stay that way and which ones need to be changed to absolute ones in order to stay consistent with those principles. Furthermore, for me it's the opposite of "obfuscating code readability": I find it harder to read. Needs constant awareness where I am and what else is in that module. To me that is in fact only relevant when refactoring and that's the point where - in those scenarios - the relative imports are a pain not a help in my view.

2.
I don't even really buy the test scenario, although I'm willing to accept it, if insisted upon. I think the entire "move the code easily" argument is flawed (in any scenario). It may be, that I can easily move `submodule/*.py` and `submodule/tests/*.py` together w/o changing imports in the latter. However, if I do that, I'll have to adjust the imports for `submodule` through the code base anyway (otherwise it's unused and should be deleted rather than moved). Whether or not it's one more file doesn't matter at all. Furthermore import adjustments for movement of a module are easiest when everything absolute, since it's one simple replacement everywhere - within and outside of the module. And finally: I don't know the realm of vi plugins, but for everyone using something like PyCharm, it's "right click, left click, type in target, enter" - if everything is absolute. If not, then I need to go through the changes done by this and find which ones of the adjustments I need to turn into relative imports again. So, to me even the relative imports in tests are causing more effort to move a module, not less.

In conclusion: I'd be more happy with absolute everywhere - no exception. However, if there remains insistence on exceptions, I'm fine with the new wording suggested and not willing to fight over it really. It's not important to me, just an annoyance. :)

Reviewer: > Whether or not it's one more file doesn't matter at all.

a minor point for defense: depending on the size/content of the file, and how move/adjustment is done (in a single commit or two commits) it would matter for `git` to identify the `mv` (AFAIK commit itself does not record that something is moved -- it is deduced "after-fact", I could be wrong), which might be crucial for seamless merging of `maint` into `master` whenever file was moved in `master. Introducing changes to the moved file in the same commit for imports (as any fancy IDE would do) should generally be "ok" since `git` allows for some "fuzz" in deducing the `mv`. Although `git` is smart that way, making it more obvious (no changes to moved file whatsoever) would assist `git` in figuring out the `mv` or might even become instrumental, if the file  is just imports and thus most of the content would be changed.

side note, unrelated to imports: for such `mv` commits, we **must not** introduce a shim in the old location in the same commit. That would make it even harder for `git` to deduce the move.  Unfortunate truth, is that ATM with `git 2.32.0` it still "does not matter" in that even with 2 commits, `git` would fail to identify that change introduced in original branch in original location, should be applied to the moved location, and not the shim.  But I am hopeful that `git` would get smarter ;)

<details>
<summary>FWIW -- demo reproducer. As long as shim is there -- `git merge` would fail :-/ </summary> 

```bash
#!/bin/bash

export PS4='> '
set -eu
cd "$(mktemp -d ${TMPDIR:-/tmp}/ann-XXXXXXX)"

set -x

git init 

mkdir sub1

cat >| sub1/mod1.py << EOF

from datalad.bugiwoogie import smth

from datalad.base import blah
from datalad.mod1.here import bleh

def test_blah():
    assert blah
EOF

git add sub1/mod1.py
git commit -m initialaddsub1

git checkout -b maint
sed -i -e 's,assert,better_assert,g' sub1/mod1.py
git commit -m 'fixed a bug in maint' -a

git checkout -
mkdir sub2
git mv sub1/* sub2/

# add change or not
# sed -i -e 's,sub1,sub2,g' sub2/*

git commit -m 'moved and adjusted sub1 content uder sub2 in master'

cat >| sub1/mod1.py << EOF
# we moved into sub1
EOF
git add sub1/mod1.py

git commit -m 'added placeholder under sub1 in master'

git merge -m 'merge maint after move' maint

```
</details>

- Imports *must* be placed at the top of a source file, unless there is a
  specific reason not to do so (e.g., delayed import due to performance
  concerns, circular dependencies). If such a reason exists, it *must*

>   should be documented by a comment at the import statement.

Author: ```suggestion
  be documented by a comment at the import statement.
```

- There *must* be no more than one import per line.

- Multiple individual imports from a single module *must* follow the pattern::

      from <module> import (
          symbol1,
          symbol2,
      )

  Individual imported symbols *should* be sorted alphabetically. The last symbol
  line *should* end with a comma.

> - Imports from packages and modules *may* be grouped in categories like

Reviewer 2: I think, while `may be grouped` is proper description of status quo, it should actually be `should be grouped` at least.
Not only does PEP8 that as well, but more importantly: There's little to no utility in doing it sometimes and sometimes not. Either it's something that one can rely on to get a quick grasp or I need to inspect every file's import closely to know.

Author: ```suggestion
- Imports from packages and modules *should* be grouped in categories like
```

  - Standard library packages

  - 3rd-party packages

>   - Datalad core
> 
>   - Datalad extensions

Reviewer: we should check how things would look like with https://github.com/PyCQA/isort 

```
  - Datalad extensions
```
is relevant only within extensions

Echoing above, may be 
```suggestion
  - Datalad core (absolute imports)

  - Datalad extensions
  
  - DataLad core ("local" relative imports)
```

Author: ```suggestion
  - Datalad core (absolute imports)

  - Datalad extensions
  
  - Datalad core ("local" relative imports)
  
  Sorting imports can be aided by https://github.com/PyCQA/isort (e.g. `python -m isort -m3 -fgw 2 -tc <filename>`).
```

.. -*- mode: rst -*-
.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:

.. _chap_design_python_imports:

************************
Python import statements
************************

.. topic:: Specification scope and status

   This specification describes the current (albeit incomplete) implementation.

> The following rules apply to any ``import`` statement in the code base:

Reviewer: brief overview of various "mainstream" libraries on use of relative imports

<details>
<summary>cpython: good amount of local, or one level up, but (thank god) not that many for higher up imports,majority of imports absolute (but didn't check if used where local could be used); tests seems to use abs imports</summary> 

```shell
$> git grep 'from \..* import' | nl | tail       
   756	Tools/c-analyzer/c_parser/preprocessor/pure.py:from ..source import (
   757	Tools/c-analyzer/c_parser/preprocessor/pure.py:from . import common as _common
   758	Tools/c-analyzer/cpython/__main__.py:from . import _analyzer, _capi, _files, _parser, REPO_ROOT
   759	Tools/c-analyzer/cpython/_analyzer.py:from . import _parser, REPO_ROOT
   760	Tools/c-analyzer/cpython/_capi.py:from ._files import iter_header_files, resolve_filename
   761	Tools/c-analyzer/cpython/_capi.py:from . import REPO_ROOT
   762	Tools/c-analyzer/cpython/_files.py:from . import REPO_ROOT, INCLUDE_DIRS, SOURCE_DIRS
   763	Tools/c-analyzer/cpython/_parser.py:from . import REPO_ROOT
   764	Tools/peg_generator/pegen/keywordgen.py:from .build import build_parser, generate_token_definitions
   765	Tools/peg_generator/pegen/keywordgen.py:from .c_generator import CParserGenerator

$> git grep 'from \.\..* import' | nl | tail
   204	Misc/HISTORY:- The compiler will report a SyntaxError if "from ... import *" occurs
   205	Misc/HISTORY:There's also a variant for 'from ... import':
   206	Misc/HISTORY:emulating from ... import *.
   207	Misc/NEWS.d/3.7.0a1.rst:Set the 'path' and 'name' attribute on ImportError for ``from ... import
   208	Misc/NEWS.d/3.8.0a1.rst:Importing names from already imported module with "from ... import ..." is
   209	Tools/c-analyzer/c_parser/parser/__init__.py:from ..info import ParsedItem
   210	Tools/c-analyzer/c_parser/parser/_info.py:from ..info import KIND, ParsedItem, FileInfo
   211	Tools/c-analyzer/c_parser/preprocessor/__init__.py:from ..source import (
   212	Tools/c-analyzer/c_parser/preprocessor/common.py:from ..info import FileInfo, SourceLine
   213	Tools/c-analyzer/c_parser/preprocessor/pure.py:from ..source import (

$> git grep 'from [^.].* import' | nl | tail 
  5705	Tools/unittestgui/unittestgui.py:from tkinter import simpledialog
  5706	setup.py:from glob import glob, escape
  5707	setup.py:    from distutils.command.build_ext import build_ext
  5708	setup.py:    from distutils.command.build_scripts import build_scripts
  5709	setup.py:    from distutils.command.install import install
  5710	setup.py:    from distutils.command.install_lib import install_lib
  5711	setup.py:    from distutils.core import Extension, setup
  5712	setup.py:    from distutils.errors import CCompilerError, DistutilsError
  5713	setup.py:    from distutils.spawn import find_executable
  5714	setup.py:        from os.path import join, exists

```
</details>

<details>
<summary>numpy: a good amount of "semantically bound/local" relative imports, tests do not use relative imports despite seems to be consistently located through the hierarchy</summary> 

```shell
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.' | nl | tail
   359	numpy/typing/_ufunc.pyi:from ._scalars import _ScalarLike_co
   360	numpy/typing/_ufunc.pyi:from ._array_like import ArrayLike, _ArrayLikeBool_co, _ArrayLikeInt_co
   361	numpy/typing/_ufunc.pyi:from ._dtype_like import DTypeLike
   362	numpy/version.py:from ._version import get_versions
   363	tools/npy_tempita/__init__.py:from ._looper import looper
   364	tools/npy_tempita/__init__.py:from .compat3 import (
   365	tools/npy_tempita/_looper.py:from .compat3 import basestring_
   366	tools/openblas_support.py:                        warnings.warn("loaded more than 1 DLL from .libs:"
   367	versioneer.py:    from ._version import get_versions
   368	versioneer.py:from ._version import get_versions
(dev3) 1 15280 [2].....................................:Wed 18 Aug 2021 12:24:37 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.\.'          
numpy/testing/_private/nosetester.py:                from ...distutils import cpuinfo
(dev3) 1 15281 [2].....................................:Wed 18 Aug 2021 12:25:00 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> find -iname tests
./numpy/core/tests
./numpy/distutils/tests
./numpy/f2py/tests
./numpy/lib/tests
./numpy/linalg/tests
./numpy/random/tests
./numpy/testing/tests
./numpy/fft/tests
./numpy/tests
./numpy/ma/tests
./numpy/compat/tests
./numpy/matrixlib/tests
./numpy/polynomial/tests
./numpy/typing/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Compiler/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Build/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/lib/python3.7/site-packages/Cython/Debugger/Tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/doc/sphinxext/numpydoc/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/compat/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/core/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/distutils/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/f2py/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/fft/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/lib/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/linalg/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/ma/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/matrixlib/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/polynomial/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/random/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/testing/tests
./benchmarks/env/cfdab1d09514b11564600fe20aac1b11/project/numpy/tests
(dev3) 1 15282 [2].....................................:Wed 18 Aug 2021 12:25:21 PM EDT:.
(git)lena:~/proj/numpy[master]git
$> git grep 'from \.' | grep tests/ | nl | tail
    14	numpy/f2py/tests/test_regression.py:from . import util
    15	numpy/f2py/tests/test_return_character.py:from . import util
    16	numpy/f2py/tests/test_return_complex.py:from . import util
    17	numpy/f2py/tests/test_return_integer.py:from . import util
    18	numpy/f2py/tests/test_return_logical.py:from . import util
    19	numpy/f2py/tests/test_return_real.py:from . import util
    20	numpy/f2py/tests/test_semicolon_split.py:from . import util
    21	numpy/f2py/tests/test_size.py:from . import util
    22	numpy/f2py/tests/test_string.py:from . import util
    23	numpy/lib/tests/test_histograms.py:            rec = sup.record(RuntimeWarning, 'Converting input from .*')
```
</details>

<details>
<summary>pandas: some amount of local imports, tests included, but majority are absolute</summary> 

```shell
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \.\.\..* import' | nl | tail
(dev3) 1 15311 [2].....................................:Wed 18 Aug 2021 12:34:12 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \..* import' | nl | tail 
   132	pandas/tests/indexes/ranges/test_range.py:from ..test_numeric import Numeric
   133	pandas/tests/indexes/timedeltas/test_timedelta.py:from ..datetimelike import DatetimeLike
   134	pandas/tests/tseries/offsets/test_fiscal.py:from .common import assert_is_on_offset, assert_offset_equal
   135	pandas/tests/tseries/offsets/test_fiscal.py:from .test_offsets import Base, WeekDay
   136	pandas/tests/tseries/offsets/test_offsets.py:from .common import assert_is_on_offset, assert_offset_equal
   137	pandas/tests/tseries/offsets/test_ticks.py:from .common import assert_offset_equal
   138	pandas/tests/tseries/offsets/test_yqm_offsets.py:from .common import assert_is_on_offset, assert_offset_equal
   139	pandas/tests/tseries/offsets/test_yqm_offsets.py:from .test_offsets import Base
   140	versioneer.py:    from ._version import get_versions
   141	versioneer.py:from ._version import get_versions
(dev3) 1 15312 [2].....................................:Wed 18 Aug 2021 12:34:17 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from \..* import' | nl | grep -v tests/ | tail
    79	pandas/_libs/tslibs/offsets.pyx:from .timedeltas import Timedelta
    80	pandas/_libs/tslibs/offsets.pyx:from .timestamps import Timestamp
    81	pandas/_libs/tslibs/offsets.pyx:        from .vectorized import normalize_i8_timestamps
    82	pandas/_libs/tslibs/tzconversion.pyx:        from .timedeltas import delta_to_nanoseconds
    83	pandas/_libs/tslibs/vectorized.pyx:from .dtypes import Resolution
    84	pandas/core/arrays/boolean.py:from .masked import BaseMaskedArray, BaseMaskedDtype
    85	pandas/core/arrays/integer.py:from .masked import BaseMaskedArray, BaseMaskedDtype
    86	pandas/core/arrays/timedeltas.py:        from .period import PeriodArray
   140	versioneer.py:    from ._version import get_versions
   141	versioneer.py:from ._version import get_versions
(dev3) 1 15313 [2].....................................:Wed 18 Aug 2021 12:34:30 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | tail             
  4547	setup.py:    from distutils.command.build_ext import build_ext as _build_ext
  4548	versioneer.py:    # keywords. When used from setup.py, we don't want to import _version.py,
  4549	versioneer.py:    # keywords. When used from setup.py, we don't want to import _version.py,
  4550	versioneer.py:from warnings import catch_warnings
  4551	versioneer.py:    from distutils.core import Command
  4552	versioneer.py:    from distutils.command.build_py import build_py as _build_py
  4553	versioneer.py:        from cx_Freeze.dist import build_exe as _build_exe
  4554	versioneer.py:        from setuptools.command.sdist import sdist as _sdist
  4555	versioneer.py:        from distutils.command.sdist import sdist as _sdist
  4556	web/pandas/community/ecosystem.md:saved/loaded to/from a file. Spyder can also import data from a variety
(dev3) 1 15314 [2].....................................:Wed 18 Aug 2021 12:34:48 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | grep tail
'   389	pandas/_libs/testing.pyx:                from pandas._testing import raise_assert_detail
   391	pandas/_libs/testing.pyx:            from pandas._testing import raise_assert_detail
   392	pandas/_libs/testing.pyx:            from pandas._testing import raise_assert_detail
  2973	pandas/tests/frame/methods/test_head_tail.py:from pandas import DataFrame
(dev3) 1 15315 [2].....................................:Wed 18 Aug 2021 12:34:54 PM EDT:.
(git)lena:~exppsy/pandas[debian]git
$> git grep 'from [^.].* import' | nl | grep 'tests/' | tail
  4473	pandas/tests/window/test_pairwise.py:from pandas import DataFrame, MultiIndex, Series, date_range
  4474	pandas/tests/window/test_pairwise.py:from pandas.core.algorithms import safe_sort
  4475	pandas/tests/window/test_rolling.py:from datetime import datetime, timedelta
  4476	pandas/tests/window/test_rolling.py:from pandas.errors import UnsupportedFunctionCall
  4477	pandas/tests/window/test_rolling.py:from pandas import DataFrame, Series, date_range
  4478	pandas/tests/window/test_rolling.py:from pandas.core.window import Rolling
  4479	pandas/tests/window/test_timeseries_window.py:from pandas import (
  4480	pandas/tests/window/test_window.py:from pandas.errors import UnsupportedFunctionCall
  4481	pandas/tests/window/test_window.py:from pandas import Series
  4482	pandas/tests/window/test_window.py:from pandas.core.window import Window

```
</details>

Author: Thanks for the empirical data! This seems to be fairly in line with the description we have in the document.

> - All imports *must* be absolute, unless they import individual pieces of an integrated code component that is only split across several source code files for technical or organizational reasons.

Reviewer: Such description reminded me about my first "large" program in Borland Pascal 7.0 on MS-DOS in 1995 where I had to learn about modularization via multiple files because the whole thing could no longer fit into maximal allowed per file 64KB ;)

But I think it is too vague and non-specific, thus provides no "rule" really. The entire `datalad` package is split into submodules for "organizational" reasons (if not at times technical such as "import time"), but we do not want to use relative imports through out.

Author: If you have a concrete suggestion on how to improve on this wording without turning it into a full page, I'd be happy to discuss.

I do not agree with the "there is no rule" judgement, though. The rule is that all imports must be absolute, unless a developer has a good reason to make an exception.

I am fairly confident that developing an explicit list of exceptions is a waste of time in a community of a handful developers, and that no list will spare us a conversation about import styles in the future. What I propose is a cheap compromise between "must be absolute", and the demands/rationals that you outlined earlier.

Reviewer: > I do not agree with the "there is no rule" judgement, though. The rule is that all imports must be absolute, unless a developer has a good reason to make an exception.

well, one developer decides it is worth exception (e.g. me on tests and "backends" etc imports) and another decides the other way -- and hence it is where we have started from really.  But I agree that this is must better than nothing due to other information in the document etc.  So indeed worthwhile proceeding as is for now.

Author: :+1:

- Imports *must* be placed at the top of a source file, unless there is a
  specific reason not to do so (e.g., delayed import due to performance
  concerns, circular dependencies). If such a reason exists, it *must*
  be documented by a comment at the import statement.

- There *must* be no more than one import per line.

- Multiple individual imports from a single module *must* follow the pattern::

      from <module> import (
          symbol1,
          symbol2,
      )

  Individual imported symbols *should* be sorted alphabetically. The last symbol
  line *should* end with a comma.

- Imports from packages and modules *should* be grouped in categories like

  - Standard library packages

  - 3rd-party packages

  - Datalad core (absolute imports)

  - Datalad extensions
  
  - Datalad core ("local" relative imports)
  

>   Sorting imports can be aided by https://github.com/PyCQA/isort (e.g. ``python -m isort -m3 -fgw 2 -tc <filename>``).

Reviewer: ref: filed https://github.com/datalad/datalad/issues/5922 on this

Author: Thx. After this PR here is merged, the instructions can be adjusted in #5922

>   Sorting imports can be aided by https://github.com/PyCQA/isort (e.g. ``python -m isort -m3 -fgw 2 -tc <filename>``).

Reviewer 3: `--fgw` and `--tc` need to be written with two hyphens.

Author: ```suggestion
  Sorting imports can be aided by https://github.com/PyCQA/isort (e.g. ``python -m isort -m3 --fgw 2 --tc <filename>``).
```

Reviewer: FWIW if #5923 is merged (I think we can do it since it should not impose anything on anyone), it could just be

```suggestion
  Sorting imports can be aided by https://github.com/PyCQA/isort (e.g. ``python -m isort <filename>``).
```
and we could (later) recommend installing/enabling `pre-commit` to automate that

Author: I'd say, let's move with this one now, and adjust later as necessary and possible.

> Example
> =======
> 
> ::
> 
>     from collections import OrderedDict
>     import logging
>     import os
> 
>     from datalad.utils import (
>         bytes2human,
>         ensure_list,
>         ensure_unicode,
>         get_dataset_root as gdr,
>     )

Reviewer: ```suggestion
Examples
========

::

    from collections import OrderedDict
    import logging
    import os

    from datalad.utils import (
        bytes2human,
        ensure_list,
        ensure_unicode,
        get_dataset_root as gdr,
    )
    
 In the `datalad/submodule/tests/test_mod.py` test file demonstrating an "exception" to absolute imports
 rule where test files are accompanying corresponding files of the underlying module:: 
 
    import os
  
    from datalad.utils import ensure_list
    
    from ..mod import func1

    from datalad.tests.utils import assert_true
    
```
